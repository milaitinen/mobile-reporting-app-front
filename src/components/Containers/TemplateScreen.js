import React, { Component } from 'react';
import {
    StyleSheet,
    View,
    FlatList,
    Platform,
    ActivityIndicator,
    ScrollView
} from 'react-native';
import { ListItem } from 'react-native-elements';

import Panel from './Panel';
import { url } from './urlsetting';

class TemplateScreen extends Component {
    static displayName = 'TemplateScreen';

    constructor(props)
    {
        super(props);
        this.state = {
            isLoading: true,
            refreshing: false,
            switch: true,
        };
    }

    componentDidMount() {

        this.getLayoutsAndForms();

    }

    getLayouts() {

        fetch(url + '/layouts')
            .then((response) => response.json())
            .then((responseJson) => {
                this.setState({
                    isLoading: false,
                    refreshing: false,
                    dataLayouts: responseJson
                }, function() {
                    // In this block you can do something with new state.
                });
            })
            .catch((error) => {
                console.error(error);
            });

    }

    getForms() {

        fetch(url + '/forms')
            .then((response) => response.json())
            .then((responseJson) => {
                this.setState({
                    isLoading: false,
                    refreshing: false,
                    dataForms: responseJson
                }, function() {
                    // In this block you can do something with new state.
                });
            })
            .catch((error) => {
                console.error(error);
            });

    }

    getLayoutsAndForms() {
        return Promise.all([this.getLayouts(), this.getForms()]);
    }


    handleRefresh = () => {
        this.setState(
            {
                refreshing: true,
            },
            () => {
                this.componentDidMount();
            }
        );

    }

    onRowPress = () => {
        this.props.navigation.navigate('NewForm', { refresh: this.handleRefresh });
    }


    render() {

        if (this.state.isLoading) {
            return (
                <View style={[styles.container]}>

                    <ActivityIndicator
                        animating={this.state.animating}
                        style={[styles.activityIndicator, { height: 80 }]}
                        size='large'
                    />

                </View>
            );
        }

        return (
            <View style={{ flex: 1 }}>

                <ScrollView contentContainerStyle={styles.MainContainer}>

                    <FlatList
                        data={ this.state.dataLayouts }
                        renderItem={({ item, index }) =>
                            <Panel
                                title={item.title}
                                onRowPress={this.onRowPress} >
                                <FlatList
                                    data={ this.state.dataForms }
                                    renderItem={({ item }) =>
                                        (item.layoutID === index + 1) ?

                                            <ListItem
                                                style={styles.FlatListItemStyle}
                                                //onPress={() => this.props.navigation.navigate('Reports')}
                                                title={item.title}
                                                subtitle={item.dateCreated}
                                            />
                                            :null

                                    }
                                    keyExtractor={item => item.orderNo}
                                />
                            </Panel>

                        }
                        keyExtractor={item => item.id}
                        refreshing={this.state.refreshing}

                    />


                </ScrollView>
            </View>
        );
    }
}



const styles = StyleSheet.create({


    activityIndicator: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        height: 80
    },

    MainContainer: {
        justifyContent: 'center',
        flex: 1,
        // margin: 10,
        paddingTop: (Platform.OS === 'ios') ? 20 : 0,
    },
    FlatListItemStyle: {
        padding: 10,
        //fontSize: 18,
        height: 60
    },
    container: {
        flex: 1,
        backgroundColor: 'white'
    }
});

export default TemplateScreen;