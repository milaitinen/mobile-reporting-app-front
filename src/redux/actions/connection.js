export const TOGGLE_CONNECTION = 'TOGGLE_CONNECTION';

export const toggleConnection = ({ connectionStatus }) => ({
    type: TOGGLE_CONNECTION,
    isConnected: connectionStatus,
});