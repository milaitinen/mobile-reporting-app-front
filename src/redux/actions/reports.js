// easier to refactor later and debug
export const STORE_REPORTS = 'STORE_REPORTS';

export const storeReports = (reports) => ({
    type: STORE_REPORTS,
    reports: reports
});
