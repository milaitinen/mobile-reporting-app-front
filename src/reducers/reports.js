import { STORE_REPORTS } from '../actions/reports';

const initialState = {
    reports: {}
};

const reportsReducer = (state = initialState, action) => {
    switch (action.type) {
        case STORE_REPORTS: {
            const newReports = action.reports.map((report) => {
                const reportObj = {};
                reportObj[report.id] = report;
                return reportObj;
            }).reduce((allReports, currentReport) => Object.assign(allReports, currentReport));

            return Object.assign(state.reports, newReports);
        }
        default:
            return state;
    }
};

export default reportsReducer;