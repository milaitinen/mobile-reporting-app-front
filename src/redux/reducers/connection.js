import { TOGGLE_CONNECTION } from '../actions/connection';

const initialState = {
    isConnected: true,
};

const connectionReducer = (state = initialState, action) => {
    switch (action.type) {
        case TOGGLE_CONNECTION:
            return {
                ...state,
                isConnected: action.isConnected,
            };
        default:
            return state;
    }
};
console.log('initialStateConnection', initialState);
export default connectionReducer;