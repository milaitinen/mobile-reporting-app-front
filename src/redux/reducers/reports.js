import { STORE_REPORTS_BY_TEMPLATE_ID, EMPTY_REPORTS } from '../actions/reports';

const initialState = {};

// return Array(Array(reports)) as object that has templateID as its keys and matching reports as its values (in Array)
const matchArrayWithTemplateID = (state, action) => {
    if (action.reports.length < 1) return {};
    return (
        action.reports.map((report) => {
            const reportsByID = {};
            // check if the array is empty
            if (report.length > 0) {
                const tempID = report[0].templateID;
                reportsByID[tempID]=report;
            }
            return reportsByID;
        }).reduce((allReports, currentReport) => Object.assign(allReports, currentReport))
    );
};

const reportsByTemplateIDReducer = (state = initialState, action) => {
    switch (action.type) {
        case STORE_REPORTS_BY_TEMPLATE_ID: {
            const newReports = matchArrayWithTemplateID(state, action);
            // Object.assign merges the given parameters together and returns an object
            return Object.assign(state.reports || {}, newReports);
        }
        case EMPTY_REPORTS: {
            return initialState;
        }
        default:
            return state;
    }
};

export default reportsByTemplateIDReducer;

