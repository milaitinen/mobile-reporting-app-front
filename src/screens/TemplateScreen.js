import React, { Component } from 'react';
import {
    View,
    FlatList,
    ActivityIndicator,
    ScrollView,
    StatusBar,
} from 'react-native';
import { connect } from 'react-redux';
import moment from 'moment';

import templateScreenStyles from './style/templateScreenStyles';
import { Layout } from '../components/Layout';
import { AppBackground } from '../components/AppBackground';
import { ReportSearchBar } from '../components/ReportSearchBar';
import { fetchReportsByTemplateID, fetchTemplatesByUsername, fetchDraftsByTemplateID } from './api';
import { storeTemplates } from '../redux/actions/templates';
import { storeReportsByTemplateID, storeDraftByTemplateID } from '../redux/actions/reports';
import { createReport } from '../redux/actions/newReport';
import { preview } from '../redux/actions/preview';
import userReducer from '../redux/reducers/user';


// "export" necessary in order to test component without Redux store
export class TemplateScreen extends Component {

    constructor(props)
    {
        super(props);
        this.state = {
            isLoading       : true,     // Checks whether the app is loading or not.
            refreshing      : false,    // Checks whether the app and its data is refreshing or not.
            scrollEnabled   : true,     // Checks whether the template screen is scrollable or not.
            renderFooter    : false     // If true, empty space is rendered after the last template. This is set to true while a template is opened.
        };
    }

    /*
    handleBackPress = () => {
        if (this.backPressed && this.backPressed > 0) {
            this.props.navigator.popToRoot({ animated: false });
            return false;
        }

        this.backPressed = 1;
        this.props.navigator.showSnackbar({
            text: 'Press one more time to exit',
            duration: 'long',
        });
        return true;
    }
    */

    /*
     componentDidMount() is invoked immediately after the component is mounted. Initialization that requires
     DOM nodes happens here. The function calls getTemplates which loads data from a remote url,
     and instantiates the network request.
    */
    componentDidMount() {
        // TEMPORARY: not sure if this is the best solution. Current version fixes a bug (related to logging in)
        if (this.props.username !== userReducer.username) {
            this.getTemplatesAndReports();
        } else {
            this.setState({ refreshing: false, isLoading: false });
        }

    }

    /*
     Fetches the data from the server in two parts.
     1) Fetches the templates from the server
     2) Fetches the reports under their specific template by making a separate fetch request using
        Promise.all. After all the promises have been fetched, the function updates the state
        of reportsByTemplates, and sets isLoading and refreshing to false.
    */
    getTemplatesAndReports = () => {
        const { username, token } = this.props;

        fetchTemplatesByUsername(username, token)
            .then(responseJson => {
                if (responseJson.length < 1) {  // handle situations where there are no templates
                    this.setState({ refreshing: false, isLoading: false });
                } else {
                    this.props.dispatch(storeTemplates(responseJson));
                }
            })
            .then(() => {
                const reportsByTemplateID = Object.keys(this.props.templates)
                    .map((templateID) => fetchReportsByTemplateID(username, templateID, token));

                Promise.all(reportsByTemplateID)
                    .then(data => this.props.dispatch(storeReportsByTemplateID(data)))
                    .then(() => this.getDrafts())
                    .catch(err => console.error(err));
            })
            .catch(error => console.error(error))
            .done();
    };

    getDrafts = () => {
        const { templates, username } = this.props;

        Object.keys(templates).forEach((templateID) => {
            fetchDraftsByTemplateID(username, templateID)
                .then((drafts) => {
                    if (drafts.length !== 0) {
                        drafts.forEach(draft => this.props.dispatch(storeDraftByTemplateID(templateID, draft)));
                    }
                })
                .then(() => this.setState({ refreshing: false, isLoading: false }))
                .catch(err => console.error(err));
        });
    };

    // Handler function for refreshing the data and refetching.
    handleRefresh = () => {
        this.setState({ refreshing: true, }, () => { this.getTemplatesAndReports(); });
        this.setState({ isLoading: true });
    };

    // Determines whether this screen is scrollable or not.
    setScrollEnabled = (bool) => {
        this.setState({ scrollEnabled : bool });
    };

    /*
     Determines whether empty space is rendered after the last template.
     Without this function it wouldn't be possible to autoscroll to the last templates.
     */
    setRenderFooter = (bool) => {
        this.setState({ renderFooter: bool });
    };

    /*
     Function that passes navigation props and navigates to NewReportScreen.
     This makes it possible for the Layout component to navigate.
     Also passes the refresh function and the specific TemplateID so that the
     app knows to which template the new report has to be added.
    */
    createNew = (templateID, isEditable) => {
        if (isEditable) {
            this.props.dispatch(createReport(templateID, moment().format('YYYY-MM-DD')));
            // this.setState({ isLoading: true });
            this.props.navigation.navigate('NewReport', { refresh: this.handleRefresh, isEditable: isEditable });
        }
        else {
            this.props.dispatch(preview(templateID));
            // this.setState({ isLoading: true });
            this.props.navigation.navigate('Preview', { refresh: this.handleRefresh,  isEditable: isEditable });
        }
        //this.setState({ isLoading: true }); TODO fix backhandler issue in NewReport, Preview, and ReportScreen and uncomment this
    };

    viewReport = (templateID, reportID, title) => {
        //this.setState({ isLoading: true }); TODO same problem as above
        this.props.navigation.navigate('Report',
            { refresh: this.handleRefresh, templateID: templateID, reportID: reportID, title: title });
    };

    render() {
        if (this.state.isLoading) {
            return (
                <AppBackground>
                    <ActivityIndicator
                        animating={this.state.animating}
                        style={[templateScreenStyles.activityIndicator, { height: 80 }]}
                        size='large'
                        color={'#88daf2'}
                    />
                </AppBackground>
            );
        }

        const { reports, templates } = this.props;

        return (
            <AppBackground>
                <View style={templateScreenStyles.viewContainer}>
                    <StatusBar
                        backgroundColor={templateScreenStyles.statusBar}
                        barStyle='light-content'
                    />
                    {/* At the moment this doesn't do anything.*/}
                    <ReportSearchBar/>
                    <ScrollView contentContainerStyle={templateScreenStyles.scrollView}>
                        <FlatList
                            ref={(c) => this._flatList = c}
                            scrollEnabled={this.state.scrollEnabled}
                            data={ Object.values(templates) }
                            renderItem={({ item, index }) =>
                                <Layout
                                    title={item.title}
                                    moveToTop={(viewPosition = 0) => this._flatList.scrollToIndex({ animated: true, index: index, viewPosition: viewPosition })}
                                    setTemplateScreenScrollEnabled={this.setScrollEnabled}
                                    setTemplateScreenRenderFooter={this.setRenderFooter}
                                    createNew={this.createNew}
                                    viewReport={this.viewReport}
                                    nofReports={(reports[item.id]) ? (reports[item.id]).length : 0}
                                    templateID={item.id}
                                    data={reports[item.id]}
                                />
                            }
                            ListFooterComponent={
                                (this.state.renderFooter) && <View style={{ height: 500 }}/>
                            }
                            keyExtractor={item => item.id}
                            refreshing={this.state.refreshing}
                        />
                    </ScrollView>
                </View>
            </AppBackground>
        );
    }
}

// maps redux state to component props. Object that is returned can be accessed via 'this.props' e.g. this.props.email
const mapStateToProps = (state) => {
    const username = state.user.username;
    const templates = state.templates;
    const reports = state.reports;
    const token = state.user.token;
    return {
        username,
        templates,
        reports,
        token
    };
};

export default connect(mapStateToProps)(TemplateScreen);
