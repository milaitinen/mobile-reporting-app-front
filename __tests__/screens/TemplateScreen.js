import 'react-native';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { ActivityIndicator } from 'react-native';
import renderer from 'react-test-renderer';

import { TemplateScreen } from '../../src/screens/TemplateScreen';
import * as connectionTypes from '../../src/redux/actions/connection';
import * as reportsTypes from '../../src/redux/actions/reports';

configure({ adapter: new Adapter() });

describe('TemplateScreen', () => {

    const dispatch = jest.fn();
    const navigate = jest.fn();
    const navigation = { navigate: navigate };
    const templates = { 2: { template_id: 2, title: 'Some template' } };
    const reports = { 2: [{ report_id: 42, title: 'Some report' }] };
    const templateScreen = shallow(<TemplateScreen templates={{}} dispatch={dispatch} navigation={navigation} />);
    const wrapper = renderer.create(
        <TemplateScreen templates={templates} reports={reports} dispatch={dispatch} navigation={navigation} />
    );


    it('renders correctly', () => {
        const tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should set state refreshing=true when handleRefresh is called', () => {

        const inst = wrapper.getInstance();

        jest.mock('../../src/api', () => {
            return {
                fetchReportsByTemplateID: jest.fn(),
                fetchTemplatesByUsername: jest.fn(),
                fetchDraftsByTemplateID: jest.fn()
            };
        });

        inst.getTemplatesAndReport = jest.fn();
        inst.handleRefresh();
        expect(inst.state.refreshing).toBe(true);
    });

    it('should set state scrollEnabled to true when called setScrollEnabled(true)', () => {
        const inst = wrapper.getInstance();

        inst.setScrollEnabled(true);
        expect(inst.state.scrollEnabled).toBe(true);
    });

    it('should set state renderFooter to false when called setRenderFooter(false)', () => {
        const inst = wrapper.getInstance();

        inst.setRenderFooter(false);
        expect(inst.state.renderFooter).toBe(false);
    });

    it('should render an ActivityIndicator if isLoading = true', () => {
        templateScreen.setState({ isLoading: true });
        expect(templateScreen.find(ActivityIndicator).length).toBe(1);
    });

    it('should not render an ActivityIndicator if isLoading = false', () => {
        templateScreen.setState({ isLoading: false });
        expect(templateScreen.find(ActivityIndicator).length).toBe(0);
    });

    it('if isEditable=false createNew() should call this.props.navigation.navigate("Preview", ...)', () => {
        const inst = wrapper.getInstance();
        inst.setState({ isNavigating: false });
        inst.createNew(1, false);

        expect(navigate).toHaveBeenCalledWith('Preview', {
            refresh: inst.handleRefresh,
            navigateDebounce: inst.handleNavigatingDebounce,
            isEditable: false
        });
    });

    it('if isEditable=true createNew() should call this.props.navigation.navigate("NewReport", ...)', () => {
        const inst = wrapper.getInstance();
        inst.setState({ isNavigating: false });
        inst.createNew(1, true);

        expect(navigate).toHaveBeenCalledWith('Report', {
            isNewReport: true,
            templateID: 1,
            navigateDebounce: inst.handleNavigatingDebounce,
            reportID: null,
            refresh: inst.handleRefresh,
            isEditable: true
        });
    });

    it('viewReport() call this.props.navigation.navigate("Report", ...)', () => {
        const inst = wrapper.getInstance();
        inst.setState({ isNavigating: false });
        inst.viewReport(1, 2, 'some title');

        expect(navigate).toHaveBeenCalledWith('Report', {
            isNewReport: false,
            refresh: inst.handleRefresh,
            navigateDebounce: inst.handleNavigatingDebounce,
            templateID: 1,
            reportID: 2,
            title: 'some title'
        });
    });

    it('dispatches toggleConnection when calling handleConnectionChange', () => {
        const inst = wrapper.getInstance();

        inst.handleConnectionChange(true);
        expect(dispatch).toHaveBeenCalledWith({ type: connectionTypes.TOGGLE_CONNECTION, isConnected: true });
    });

    it('getDraft', async () => {
        jest.unmock('../../src/api');
        const api = require.requireActual('../../src/api');
        api.fetchDraftsByTemplateID = jest.fn(() => new Promise(resolve => resolve([{ report_id: -1, title: 'draft1' }])));

        const inst = wrapper.getInstance();
        await inst.getDrafts();
        expect(dispatch).toHaveBeenCalledWith({
            type: reportsTypes.STORE_DRAFT_BY_TEMPLATE_ID,
            templateID: 2,
            draft: { report_id: -1, title: 'draft1' }
        });
    });
});
