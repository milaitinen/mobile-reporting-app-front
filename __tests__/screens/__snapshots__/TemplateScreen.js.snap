// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateScreen renders correctly 1`] = `
<View
  colors={
    Array [
      undefined,
      undefined,
      undefined,
    ]
  }
  style={undefined}
>
  <View
    style={undefined}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#393e42",
            "borderBottomColor": "#000",
            "borderBottomWidth": 1,
            "borderTopColor": "#000",
            "borderTopWidth": 1,
          },
          Object {
            "backgroundColor": "#e1e8ee",
            "borderBottomColor": "#e1e1e1",
            "borderTopColor": "#e1e1e1",
          },
          undefined,
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        placeholder="templates.search"
        placeholderTextColor="#86939e"
        style={
          Array [
            Object {
              "backgroundColor": "#303337",
              "borderRadius": 3,
              "color": "#86939e",
              "fontSize": 17.5,
              "height": 30,
              "margin": 8,
              "overflow": "hidden",
              "paddingLeft": 26,
              "paddingRight": 19,
            },
            Object {
              "backgroundColor": "#bdc6cf",
            },
            false,
            false,
            undefined,
            undefined,
            false,
          ]
        }
        underlineColorAndroid="transparent"
      />
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 16,
            },
            Array [
              Object {
                "backgroundColor": "transparent",
                "position": "absolute",
                "top": 15.5,
              },
              Object {
                "left": 16,
              },
              undefined,
            ],
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <RCTScrollView
      contentContainerStyle={undefined}
    >
      <View>
        <RCTScrollView
          ListFooterComponent={false}
          data={
            Array [
              Object {
                "template_id": 2,
                "title": "Some template",
              },
            ]
          }
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          refreshing={false}
          renderItem={[Function]}
          scrollEnabled={true}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <View
                collapsable={undefined}
                style={
                  Object {
                    "height": 60,
                  }
                }
              >
                <View>
                  <View
                    accessibilityComponentType={undefined}
                    accessibilityLabel={undefined}
                    accessibilityTraits={undefined}
                    accessible={true}
                    hasTVPreferredFocus={undefined}
                    hitSlop={undefined}
                    isTVSelectable={true}
                    nativeID={undefined}
                    onLayout={undefined}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "borderBottomColor": "#bbb",
                            "borderBottomWidth": 1,
                            "paddingBottom": 10,
                            "paddingRight": 10,
                            "paddingTop": 10,
                          },
                          undefined,
                          false,
                          false,
                        ],
                      ]
                    }
                    testID={undefined}
                    tvParallaxProperties={undefined}
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "flexDirection": "row",
                            "marginLeft": 10,
                          },
                          undefined,
                        ]
                      }
                    >
                      <View
                        disabled={false}
                        onLongPress={undefined}
                        onPress={undefined}
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "justifyContent": "center",
                            },
                            undefined,
                          ]
                        }
                        underlayColor="white"
                      >
                        <View>
                          <View
                            onPress={undefined}
                            style={
                              Array [
                                false,
                                false,
                                false,
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "transparent",
                                  "justifyContent": "center",
                                },
                                undefined,
                              ]
                            }
                            underlayColor="white"
                          >
                            <Text
                              accessible={true}
                              allowFontScaling={false}
                              ellipsizeMode="tail"
                              style={
                                Array [
                                  Object {
                                    "color": "#bdc6cf",
                                    "fontSize": 24,
                                  },
                                  Array [
                                    Object {
                                      "backgroundColor": "transparent",
                                    },
                                    Array [
                                      Object {
                                        "marginRight": 8,
                                      },
                                      undefined,
                                    ],
                                  ],
                                  Object {
                                    "fontFamily": "Material Icons",
                                    "fontStyle": "normal",
                                    "fontWeight": "normal",
                                  },
                                ]
                              }
                            >
                              
                            </Text>
                          </View>
                        </View>
                      </View>
                      <View
                        style={
                          Object {
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <View
                          style={undefined}
                        >
                          <Text
                            accessible={true}
                            allowFontScaling={true}
                            ellipsizeMode="tail"
                            numberOfLines={1}
                            style={
                              Array [
                                Object {},
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                                Array [
                                  Object {
                                    "color": "#43484d",
                                    "fontSize": 17.5,
                                  },
                                  false,
                                  undefined,
                                  undefined,
                                ],
                              ]
                            }
                          >
                            Some template
                          </Text>
                        </View>
                        <View
                          style={undefined}
                        >
                          <View>
                            <View
                              style={undefined}
                            >
                              <Text
                                accessible={true}
                                allowFontScaling={true}
                                ellipsizeMode="tail"
                              >
                                1
                                 
                                templates.report
                              </Text>
                              <Text
                                accessible={true}
                                allowFontScaling={true}
                                ellipsizeMode="tail"
                              >
                                0
                                 
                                templates.queuedPlural
                              </Text>
                            </View>
                          </View>
                        </View>
                      </View>
                      <View
                        style={undefined}
                      >
                        <View
                          accessibilityComponentType={undefined}
                          accessibilityLabel={undefined}
                          accessibilityTraits={undefined}
                          accessible={true}
                          hasTVPreferredFocus={undefined}
                          hitSlop={undefined}
                          isTVSelectable={true}
                          nativeID={undefined}
                          onLayout={undefined}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                              },
                              undefined,
                            ]
                          }
                          testID={undefined}
                          tvParallaxProperties={undefined}
                        >
                          <View>
                            <View
                              onPress={undefined}
                              style={
                                Array [
                                  false,
                                  false,
                                  false,
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "transparent",
                                    "justifyContent": "center",
                                  },
                                  undefined,
                                ]
                              }
                              underlayColor="white"
                            >
                              <Text
                                accessible={true}
                                allowFontScaling={false}
                                ellipsizeMode="tail"
                                style={
                                  Array [
                                    Object {
                                      "color": "black",
                                      "fontSize": 24,
                                    },
                                    Array [
                                      Object {
                                        "backgroundColor": "transparent",
                                      },
                                      Array [
                                        undefined,
                                        Object {
                                          "backgroundColor": "#A4CBE7",
                                          "color": "#8096a8",
                                        },
                                      ],
                                    ],
                                    Object {
                                      "fontFamily": "Material Icons",
                                      "fontStyle": "normal",
                                      "fontWeight": "normal",
                                    },
                                  ]
                                }
                              >
                                
                              </Text>
                            </View>
                          </View>
                        </View>
                        <View
                          accessibilityComponentType={undefined}
                          accessibilityLabel={undefined}
                          accessibilityTraits={undefined}
                          accessible={true}
                          hasTVPreferredFocus={undefined}
                          hitSlop={undefined}
                          isTVSelectable={true}
                          nativeID={undefined}
                          onLayout={undefined}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                              },
                              undefined,
                            ]
                          }
                          testID={undefined}
                          tvParallaxProperties={undefined}
                        >
                          <View>
                            <View
                              onPress={undefined}
                              style={
                                Array [
                                  false,
                                  false,
                                  false,
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "transparent",
                                    "justifyContent": "center",
                                  },
                                  undefined,
                                ]
                              }
                              underlayColor="white"
                            >
                              <Text
                                accessible={true}
                                allowFontScaling={false}
                                ellipsizeMode="tail"
                                style={
                                  Array [
                                    Object {
                                      "color": "black",
                                      "fontSize": 24,
                                    },
                                    Array [
                                      Object {
                                        "backgroundColor": "transparent",
                                      },
                                      Array [
                                        undefined,
                                        Object {
                                          "backgroundColor": "#99d9ad",
                                          "color": "#77a482",
                                        },
                                      ],
                                    ],
                                    Object {
                                      "fontFamily": "Material Icons",
                                      "fontStyle": "normal",
                                      "fontWeight": "normal",
                                    },
                                  ]
                                }
                              >
                                
                              </Text>
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
                <View
                  onLayout={[Function]}
                  style={undefined}
                >
                  <RCTScrollView
                    style={
                      Object {
                        "height": 415,
                      }
                    }
                  >
                    <View>
                      <View
                        ListFooterComponent={
                          <Text
                            accessible={true}
                            allowFontScaling={true}
                            ellipsizeMode="tail"
                            style={undefined}
                          >
                            templates.endOfReports
                          </Text>
                        }
                        data={
                          Array [
                            Object {
                              "report_id": 42,
                              "title": "Some report",
                            },
                          ]
                        }
                        disableVirtualization={false}
                        extraData={20}
                        getItem={[Function]}
                        getItemCount={[Function]}
                        horizontal={false}
                        initialNumToRender={10}
                        keyExtractor={[Function]}
                        maxToRenderPerBatch={10}
                        numColumns={1}
                        onContentSizeChange={[Function]}
                        onEndReachedThreshold={2}
                        onLayout={[Function]}
                        onMomentumScrollEnd={[Function]}
                        onScroll={[Function]}
                        onScrollBeginDrag={[Function]}
                        onScrollEndDrag={[Function]}
                        renderItem={[Function]}
                        scrollEventThrottle={50}
                        stickyHeaderIndices={Array []}
                        updateCellsBatchingPeriod={50}
                        viewabilityConfigCallbackPairs={Array []}
                        windowSize={21}
                      >
                        <View
                          onLayout={[Function]}
                          style={null}
                        >
                          <View
                            accessibilityComponentType={undefined}
                            accessibilityLabel={undefined}
                            accessibilityTraits={undefined}
                            accessible={true}
                            hasTVPreferredFocus={undefined}
                            hitSlop={undefined}
                            isTVSelectable={true}
                            nativeID={undefined}
                            onLayout={undefined}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "backgroundColor": "transparent",
                                },
                                Array [
                                  Object {
                                    "backgroundColor": "transparent",
                                    "borderBottomColor": "#bbb",
                                    "borderBottomWidth": 1,
                                    "paddingBottom": 10,
                                    "paddingRight": 10,
                                    "paddingTop": 10,
                                  },
                                  undefined,
                                  false,
                                  false,
                                ],
                              ]
                            }
                            testID={undefined}
                            tvParallaxProperties={undefined}
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "flexDirection": "row",
                                    "marginLeft": 10,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                style={
                                  Object {
                                    "flex": 1,
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                <View
                                  style={undefined}
                                >
                                  <Text
                                    accessible={true}
                                    allowFontScaling={true}
                                    ellipsizeMode="tail"
                                    numberOfLines={1}
                                    style={
                                      Array [
                                        Object {},
                                        undefined,
                                        undefined,
                                        undefined,
                                        undefined,
                                        undefined,
                                        undefined,
                                        undefined,
                                        undefined,
                                        undefined,
                                        Array [
                                          Object {
                                            "color": "#43484d",
                                            "fontSize": 17.5,
                                          },
                                          Object {
                                            "marginLeft": 10,
                                          },
                                          undefined,
                                          undefined,
                                        ],
                                      ]
                                    }
                                  >
                                    Some report
                                  </Text>
                                </View>
                                <View
                                  style={undefined}
                                >
                                  <View />
                                </View>
                              </View>
                              <View
                                style={undefined}
                              >
                                <View
                                  style={
                                    Array [
                                      undefined,
                                    ]
                                  }
                                >
                                  <View
                                    onPress={undefined}
                                    style={
                                      Array [
                                        Object {
                                          "alignItems": "center",
                                          "backgroundColor": "#43484d",
                                          "borderRadius": 20,
                                          "justifyContent": "center",
                                          "padding": 12,
                                          "paddingBottom": 3,
                                          "paddingTop": 3,
                                        },
                                        Array [
                                          undefined,
                                          Object {
                                            "backgroundColor": "#f3fe99",
                                          },
                                        ],
                                      ]
                                    }
                                  >
                                    <Text
                                      accessible={true}
                                      allowFontScaling={true}
                                      ellipsizeMode="tail"
                                      style={undefined}
                                    >
                                      templates.sent
                                    </Text>
                                    <View
                                      onPress={undefined}
                                      style={
                                        Array [
                                          false,
                                          false,
                                          false,
                                          Object {
                                            "alignItems": "center",
                                            "backgroundColor": "transparent",
                                            "justifyContent": "center",
                                          },
                                          undefined,
                                        ]
                                      }
                                      underlayColor="white"
                                    >
                                      <Text
                                        accessible={true}
                                        allowFontScaling={false}
                                        ellipsizeMode="tail"
                                        style={
                                          Array [
                                            Object {
                                              "color": "black",
                                              "fontSize": 24,
                                            },
                                            Array [
                                              Object {
                                                "backgroundColor": "transparent",
                                              },
                                              undefined,
                                            ],
                                            Object {
                                              "fontFamily": "Feather",
                                              "fontStyle": "normal",
                                              "fontWeight": "normal",
                                            },
                                          ]
                                        }
                                      >
                                        
                                      </Text>
                                    </View>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                        <View
                          onLayout={[Function]}
                          style={null}
                        >
                          <Text
                            accessible={true}
                            allowFontScaling={true}
                            ellipsizeMode="tail"
                            style={undefined}
                          >
                            templates.endOfReports
                          </Text>
                        </View>
                      </View>
                    </View>
                  </RCTScrollView>
                </View>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
